manifests/init.pp

```diff
@@ -48,94 +42,103 @@
+  String[1]                           $desktop_package_name  = 'open-vm-tools-desktop',
+  Boolean                             $manage_epel           = false,
+  String[1]                           $package_name          = 'open-vm-tools',
+  Stdlib::Ensure::Service             $service_ensure        = 'running',
+  Variant[String[1],Array[String[1]]] $service_name          = ['vgauthd','vmtoolsd'],
+  Boolean                             $service_enable        = true,
+  Boolean                             $service_hasstatus     = true,
+  String[1]                           $service_pattern       = 'vmtoolsd',
+  Boolean                             $with_desktop          = false,
+) {
 
-  $supported = $openvmtools::params::supported
+  $supported = lookup('openvmtools::supported', Boolean, 'first', false)
 
-  # Validate our booleans
-  validate_bool($with_desktop)
-  validate_bool($autoupgrade)
-  validate_bool($service_enable)
-  validate_bool($service_hasstatus)
-  validate_bool($supported)
 
-  case $ensure {
-    /(present)/: {
-      if $autoupgrade == true {
-        $package_ensure = 'latest'
-      } else {
-        $package_ensure = 'present'
-      }
-
-      if $service_ensure in [ running, stopped ] {
-        $service_ensure_real = $service_ensure
-      } else {
-        fail('service_ensure parameter must be running or stopped')
-      }
-    }
-    /(absent)/: {
-      $package_ensure = 'absent'
-      $service_ensure_real = 'stopped'
-    }
-    default: {
-      fail('ensure parameter must be present or absent')
+  if $ensure == 'present' {
+    $package_ensure = $autoupgrade ? {
+      true    => 'latest',
+      default => 'present',
     }
+    $service_ensure_real = $service_ensure
+  }
+  else {  # ensure == 'absent'
+    $package_ensure = 'absent'
+    $service_ensure_real = 'stopped'
   }
 
-  case $::virtual {
-    'vmware': {
-      if $supported {
-        package { $package_name :
-          ensure => $package_ensure,
-        }
-        if $with_desktop {
-          package { $desktop_package_name :
-            ensure => $package_ensure,
-          }
-        }
-
-        service { $service_name :
-          ensure    => $service_ensure_real,
-          enable    => $service_enable,
-          hasstatus => $service_hasstatus,
-          pattern   => $service_pattern,
-          require   => Package[$package_name],
+  if defined('$facts["virtual"]') && $facts['virtual'] == 'vmware' {
+    if $supported {
+      $packages = $with_desktop ? {
+        true    => $desktop_package_conflicts ? {
+          true    => [ $desktop_package_name ],
+          default => [ $package_name, $desktop_package_name ],
+        },
+        default => [ $package_name ],
+      }   
+      if $manage_epel {
+        include epel
+        Yumrepo['epel'] -> Package[$packages]
+      }
+      $vmware_uninstaller = defined('$facts["vmware_tools_uninstaller_path"]') ? {
```

Comment:
This `defined()` check is redundant.

You could simplify all this and combine the below uninstaller bit of code with:

```puppet
if $facts['vmware_tools_uninstaller_path'] {
  $vmware_lib = $vmware_uninstaller.regex_replace('bin/vmware-uninstall-tools.pl','lib/vmware-tools')

  exec { 'vmware-uninstall-tools':
    command => "${vmware_uninstaller} && rm -rf ${vmware_lib}",
    before  => Package['VMwareTools'],
  }
}

Reply:
What happens if I try to reference a fact that hasn't been defined?  Doesn't it throw an error?

Reply:
Nope. It's value is just `undef`, which is falsey.

```
[root@agent ~]# cat variable.pp
if $facts['this_fact_does_not_exist'] {
  notice('you will not see this')
}

[root@agent ~]# puppet apply variable.pp
Notice: Compiled catalog for agent.vagrant in environment production in 0.07 seconds
Notice: Applied catalog in 0.10 seconds

[root@agent ~]# puppet --version
4.10.12
```

Reply:
Also, I still have to assign `$vmware_uninstaller`, or else use the `$fact` each time.

Reply:
I'll just shorten the fact-name.

Reply:
Ah right. I missed that bit... the point still stands though, you don't need `defined()`.

```
if $facts['vmware_tools_uninstaller_path'] {
  $vmware_lib = $facts['vmware_tools_uninstaller_path'].regex_replace('bin/vmware-uninstall-tools.pl','lib/vmware-tools')

  exec { 'vmware-uninstall-tools':
    command => "${vmware_uninstaller} && rm -rf ${vmware_lib}",
    before  => Package['VMwareTools'],
  }
}
```